File: destack/core/definition/schema.py
from typing import final

from ..builtin import StructType, declare_property, declare_struct
from .definition import Definition
from .enum import EnumDefinition
from .handle import HandleDefinition
from .module import ModuleDefinition
from .node import NodeDefinition
from .struct import StructDefinition


@declare_struct(
    StructType.SCHEMA_DEFINITION,
    is_final=True,
)
@final
class SchemaDefinition(Definition):
    """Definition of the entire Destack Schema ("language definition")."""

    # meta
    version: str = declare_property(108)

    # content
    modules: list[ModuleDefinition] = declare_property(120)
    nodes: list[NodeDefinition] = declare_property(121)
    structs: list[StructDefinition] = declare_property(122)
    handles: list[HandleDefinition] = declare_property(123)
    enums: list[EnumDefinition] = declare_property(124)

File: destack-rs/destack/src/core/definition/schema.rs
//! destack.core.definition.schema

#![destack::partial(destack.core.definition.schema, file)]

use crate::{EnumDefinition, HandleDefinition, ModuleDefinition, NodeDefinition, StructDefinition};

#[destack::generated(SchemaDefinition, -, block)]
/// Definition of the entire Destack Schema ("language definition").
pub struct SchemaDefinition {
    pub name: String,
    pub description: String,
    pub version: String,
    pub modules: Vec<ModuleDefinition>,
    pub nodes: Vec<NodeDefinition>,
    pub structs: Vec<StructDefinition>,
    pub handles: Vec<HandleDefinition>,
    pub enums: Vec<EnumDefinition>,
}
